### PR TITLE
Added Villager Trades and basic movement

### PR DESCRIPTION
**Pull request by Buddelbubi**

### CODE REVIEW

**Comment by smartcmd** [comment]
Cool :heart:

File: pom.xml [hunk 1]
```diff
@@ -543,6 +537,12 @@
             <version>1.1.10.1</version>
             <scope>compile</scope>
         </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <version>1.18.22</version>
```
Comment (SuperIceCN) [inline pom.xml:543]: Why do we need to roll back the version from `1.18.24` to `1.18.22`?

File: src/main/java/cn/nukkit/entity/passive/EntityVillager.java [hunk 2]
```diff
@@ -76,60 +98,54 @@ public class EntityVillager extends EntityCreature implements InventoryHolder, E
...
+                                all(
+                                        new PassByTimeEvaluator(CoreMemoryTypes.ENTITY_SPAWN_TIME, 20 * 60 * 20, Integer.MAX_VALUE),
+                                        entity -> entity instanceof EntityAnimal animal && animal.isBaby()
+                                )
+                                , 1, 1, 1200
+                        )
+                ),
+                Set.of(
+                        new Behavior(new FlatRandomRoamExecutor(0.4f, 12, 40, true, 100, true, 10), new PassByTimeEvaluator(CoreMemoryTypes.LAST_BE_ATTACKED_TIME, 0, 100), 4, 1),
+                        new Behavior(new EntityBreedingExecutor<>(EntityPig.class, 16, 100, 0.5f), entity -> entity.getMemoryStorage().get(CoreMemoryTypes.IS_IN_LOVE), 3, 1),
```
Comment (SuperIceCN) [inline src/main/java/cn/nukkit/entity/passive/EntityVillager.java:136]: Why do we need `EntityBreedingExecutor` with `EntityPig.class`? Can we feed the villagers?

File: src/main/java/cn/nukkit/entity/passive/EntityVillager.java [hunk 3]
```diff
@@ -76,60 +98,54 @@ public class EntityVillager extends EntityCreature implements InventoryHolder, E
...
+                        new Behavior(new MoveToTargetExecutor(CoreMemoryTypes.NEAREST_FEEDING_PLAYER, 0.4f, true), new MemoryCheckNotEmptyEvaluator(CoreMemoryTypes.NEAREST_FEEDING_PLAYER), 2, 1),
+                        new Behavior(new LookAtTargetExecutor(CoreMemoryTypes.NEAREST_PLAYER, 100), new ProbabilityEvaluator(4, 10), 1, 1, 100),
+                        new Behavior(new FlatRandomRoamExecutor(0.2f, 12, 100, false, -1, true, 10), (entity -> true), 1, 1)
+                ),
+                Set.of(new NearestFeedingPlayerSensor(8, 0), new NearestPlayerSensor(8, 0, 20)),
```
Comment (SuperIceCN) [inline src/main/java/cn/nukkit/entity/passive/EntityVillager.java:141]: If we use NearestFeedingPlayerSensor, the Entity must extends EntityAnimal. However it's not the case for villager now.

File: src/main/java/cn/nukkit/entity/passive/EntityVillager.java [hunk 4]
```diff
@@ -385,4 +405,90 @@ public boolean onInteract(Player player, Item item, Vector3 clickedPos) {
...
+                        Block block = getLocation().add(x, 0, z).getLevelBlock();
+                        int id = block.getId();
+                        for (Profession profession : Profession.getProfessions().values()) {
+                            if (id == profession.getBlockid()) {
+                                professionFound = true;
+                                if (this.profession != profession.getIndex()) {
+                                    this.setTradeSeed(new Random().nextInt(Integer.MAX_VALUE));
+                                    this.setProfession(profession.getIndex());
+                                    this.applyProfession(profession);
+                                    BlockVector3 blockVector = block.getLocation().asBlockVector3();
```
Comment (SuperIceCN) [inline src/main/java/cn/nukkit/entity/passive/EntityVillager.java:459]: Using block.getFloorX/Y/Z instead.

File: src/main/java/cn/nukkit/entity/passive/EntityVillager.java [hunk 5]
```diff
@@ -385,4 +405,90 @@ public boolean onInteract(Player player, Item item, Vector3 clickedPos) {
...
+                boolean professionFound = false;
+                for(int x = -1; x <= 1; x++) {
+                    for (int z = -1; z <= 1; z++) {
+                        Block block = getLocation().add(x, 0, z).getLevelBlock();
+                        int id = block.getId();
+                        for (Profession profession : Profession.getProfessions().values()) {
+                            if (id == profession.getBlockid()) {
+                                professionFound = true;
+                                if (this.profession != profession.getIndex()) {
+                                    this.setTradeSeed(new Random().nextInt(Integer.MAX_VALUE));
```
Comment (SuperIceCN) [inline src/main/java/cn/nukkit/entity/passive/EntityVillager.java:456]: Using ThreadLocalRandom or NukkitRandom from the level of this villager instead, Random is globally synchronized and may block PNX, thereby affecting performance.

**Review by SuperIceCN — CHANGES_REQUESTED** [review]
Also, please format your code:

1. Remove the unused imports
2. There should be space between expression tokens, e.g. `tick % (10 * 10) == 0`
3. There should be space between if/for/while, conditions and code blocks
4. No empty lines at the start and the end of methods.

Anyway, thank you very much for your contribution to PNX!

File: pom.xml (hunk 1 above)
Comment (Buddelbubi) [reply pom.xml:543]: IntelliJ did that

Comment (Buddelbubi) [reply pom.xml:543]: You dont

File: src/main/java/cn/nukkit/entity/passive/EntityVillager.java (hunk 2 above)
Comment (Buddelbubi) [reply src/main/java/cn/nukkit/entity/passive/EntityVillager.java:136]: Uh. I focused on the functionality. I probably missed that

**Review by CoolLoong — CHANGES_REQUESTED** [review]

**Comment by CoolLoong** [comment]
@Buddelbubi You did not commit cn.nukkit.entity.data.profession.Profession

**Comment by SuperIceCN** [comment]
May I ask if you have time to resolve the problems in this PR recently?

**Comment by Buddelbubi** [comment]
I already did. Did I miss some?

**Comment by CoolLoong** [comment]
thank you for your PR

File: src/main/java/cn/nukkit/network/protocol/ProtocolInfo.java [hunk 6]
```diff
@@ -18,7 +18,7 @@ public interface ProtocolInfo {
     /**
      * Actual Minecraft: PE protocol version
      */
-    int CURRENT_PROTOCOL = dynamic(589);
+    int CURRENT_PROTOCOL = dynamic(594);
```
Comment (CoolLoong) [inline src/main/java/cn/nukkit/network/protocol/ProtocolInfo.java:21]: please wait https://github.com/PowerNukkitX/PowerNukkitX/pull/1312

**Review by CoolLoong — CHANGES_REQUESTED** [review]

**Comment by CoolLoong** [comment]
Other than that, there are no issue

**Review by CoolLoong — APPROVED** [review]

**Comment by CoolLoong** [comment]
@SuperIceCN

**Review by smartcmd — APPROVED** [review]

**Review by SuperIceCN — APPROVED** [review]

**Pull request merged**